x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/StreamingHttpResultPublisher.java

```diff
@@ -0,0 +1,232 @@
+                    var maxBytes = settings.getMaxResponseSize().getBytes() * 0.5;
+                    if (bytesInQueue.updateAndGet(current -> Long.min(0, current - allBytes.length)) <= maxBytes) {
+                        resumeProducer();
+                    }
+                }
+            });
+        }
+
+        // always check if totalByteSize > the configured setting in case the settings change
+        if (bytesInQueue.accumulateAndGet(allBytes.length, Long::sum) >= settings.getMaxResponseSize().getBytes()) {
+            pauseProducer(ioControl);
+        }
+
+        // always run in case we're waking up from a pause and need to start a new thread
+        taskRunner.requestNextRun();
+    }
+
+    private void pauseProducer(IOControl ioControl) {
+        ioControl.suspendInput();
+        synchronized (ioLock) {
+            savedIoControl = ioControl;
+        }
+    }
+
+    private void resumeProducer() {
+        synchronized (ioLock) {
+            if (savedIoControl != null) {
+                savedIoControl.requestInput();
+                savedIoControl = null;
+            }
+        }
+    }
+
+    @Override
+    public void responseCompleted(HttpContext httpContext) {}
+
+    // called when Apache is failing the response
+    @Override
+    public void failed(Exception e) {
+        if (this.isDone.compareAndSet(false, true)) {
+            ex = e;
+            queue.offer(() -> subscriber.onError(e));
+            taskRunner.requestNextRun();
+        }
+    }
+
+    // called when Apache is done with the response
+    @Override
+    public void close() {
+        if (isDone.compareAndSet(false, true)) {
+            queue.offer(() -> subscriber.onComplete());
+            taskRunner.requestNextRun();
+        }
+    }
+
+    // called when Apache is canceling the response
+    @Override
+    public boolean cancel() {
+        close();
+        return true;
+    }
+
+    @Override
+    public Exception getException() {
+        return ex;
+    }
+
+    @Override
+    public HttpResponse getResult() {
+        return response;
+    }
+
+    @Override
+    public boolean isDone() {
+        return isDone.get();
+    }
+
+    private class HttpSubscription implements Flow.Subscription {
+        @Override
+        public void request(long n) {
+            if (subscriptionCanceled.get()) {
+                return;
+            }
+
+            if (n > 0) {
+                pendingRequest.set(true);
+                taskRunner.requestNextRun();
+            } else {
+                // per Subscription's spec, fail the subscriber and stop the processor
+                cancel();
+                subscriber.onError(new IllegalArgumentException("Subscriber requested a non-positive number " + n));
+            }
+        }
+
+        @Override
+        public void cancel() {
+            if (subscriptionCanceled.compareAndSet(false, true)) {
+                FutureUtils.cancel(threadRunner.get());
```

Comment:
I don't see any other use of `threadRunner` do you mean `taskRunner.cancel();` ?
```